src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -811,17 +823,27 @@ public bool ApplyNullableTransforms(ImmutableArray<bool> transforms, INonNullTyp
                 result = result.WithTypeAndModifiers(newTypeSymbol, result.CustomModifiers);
             }
 
-            if (isAnnotated)
-            {
-                result = result.AsNullableReferenceType(fromDeclaration: true);
-            }
-            else if (nonNullTypesContext.NonNullTypes == true)
+            switch ((NullableAnnotation)transformFlag)
             {
-                result = result.AsNotNullableReferenceType();
-            }
-            else if (result.NullableAnnotation != NullableAnnotation.Unknown && (!result.NullableAnnotation.IsAnyNullable() || !oldTypeSymbol.IsValueType))
-            {
-                result = CreateNonLazyType(newTypeSymbol, NullableAnnotation.Unknown, result.CustomModifiers);
+                case NullableAnnotation.Nullable:
+                    result = result.AsNullableReferenceType();
+                    break;
+
+                case NullableAnnotation.NotNullable:
+                    result = result.AsNotNullableReferenceType();
+                    break;
+
+                case NullableAnnotation.Unknown:
+                    Debug.Assert((NullableAnnotation)transformFlag == NullableAnnotation.Unknown);
+                    if (result.NullableAnnotation != NullableAnnotation.Unknown && (!result.NullableAnnotation.IsAnyNullable() || !oldTypeSymbol.IsNullableType()))
```

Comment:
>(!result.NullableAnnotation.IsAnyNullable() || !oldTypeSymbol.IsNullableType()) [](start = 83, length = 79)

nit: took me a while to understand this check. Inverting this last block helps: ` .. && !(result.NullableAnnotation.IsAnyNullable() && oldTypeSymbol.IsNullableType())` 
Maybe add a comment too: "// keep nullable annotation on `Nullable<T>` types"